### PR TITLE
Refactor EndpointInfo and ConnectionInfo in C++

### PR DESCRIPTION
**Pull request by bernardnormier**
This is the C++ equivalent of #3185 and #3187.

### CODE REVIEW

File: cpp/src/Ice/SSL/SecureTransportTransceiverI.cpp [hunk 1]
```diff
@@ -503,24 +503,22 @@ Ice::SSL::SecureTransport::TransceiverI::toDetailedString() const
...
         CFRetain(peerCertificate);
-        info->peerCertificate = peerCertificate;
-    }
-    else
-    {
-        info->peerCertificate = nullptr;
     }
 
-    return info;
+    return make_shared<Ice::SSL::ConnectionInfo>(
```
Comment (pepone) [inline cpp/src/Ice/SSL/SecureTransportTransceiverI.cpp:519]: A minor issue with this update, and other certificates. If the constructor throws, or `_delegate->getInfo()` throws. we leak the peerCertificate.

**Review by pepone — APPROVED** [review]
Looks good, there was a failure with the MATLAB build.

File: cpp/src/Ice/SSL/SecureTransportTransceiverI.cpp (hunk 1 above)
Comment (bernardnormier) [reply cpp/src/Ice/SSL/SecureTransportTransceiverI.cpp:519]: It's not clear to me why the constructor or _delegate->getInfo would throw.

**Comment by bernardnormier** [comment]
> Looks good, there was a failure with the MATLAB build.

Small issue in the MATLAB code, now fixed. A more interesting issue was the failure on macos release, apparently due to weak vtables. Also fixed.

**Pull request merged**